rfc/001-composable-pipelines.md

```diff
@@ -0,0 +1,205 @@
+# Composable Pipelines
+
+- Authors:
+  - [Matthias Vallentin](https://github.com/mavam)
+
+## Overview
+
+This proposal ideates a new UX around VAST's pipelines. The fundamental idea is
+that *everything is a pipeline* and build a composable UX around this concept.
+
+## Problem Statement
+
+Pipelines currently are only deployable at [predefined
+points](https://vast.io/docs/use-vast/transform). For example, they can be
+triggered for specific events on import/export, or invoked by compaction.
+
+As we are bringing pipelines to the query language, we need to solve a UX
+challenge: how do we make the data flow explicit and provide a unified
+experience, especially in the presence of custom plugins?
+
+## Solution Proposal
+
+Assume for a moment everything is a pipeline operator. Then we may write:
+
+ ```bash
+vast 'from s3://aws |
+      read json |
+      group-by src |
+      summarize count(dst) |
+      write feather |
+      to /path/to/file.feather'
+```
+
+Just to illustrate the composability, we could take this to an extreme when
+performance doesn't matter:
+
+```bash
```

Comment:
I'm in full agreement here. I've used this disentangled example only to tease the idea of "everything is an operator" and how it enables composable building blocks.

I'm not sure whether that's a good idea in practice. For the reasons you listed, but also performance. Unless you're doing exploratory data analysis on your shell for highly selective data points, there is no good reason to do it this way.

Reply:
On the other end, I think this may make a more expressive Python API which would still be close enough to the CLI:

```bash
vast from s3://aws |
  vast read json |
  vast group-by |
  vast summarize |
  vast write feather |
  vast to /path/to/file.feather
```

```python
await vast.from('s3://aws')
          .read_json()
          .group_by()
          .summary()
          .write_feather()
          .to('/path/to/file.feather')
```

Reply:
These sort of python APIs are indeed the desired outcome. Once we have the data in Apache Arrow, we'd like to pass the execution to other more powerful data frame libraries, like [polars](https://www.pola.rs/). Here's an example, how this looks today:

```python
import polars as pl

q = (
    pl.scan_csv("iris.csv")
    .filter(pl.col("sepal_length") > 5)
    .groupby("species")
    .agg(pl.all().sum())
)

df = q.collect()
```

Likewise, we're closely following [ibis](https://ibis-project.org/).

Most of these Python DSLs build a DAG of the execution lazily and then submit it to some executor.

Reply:
(I added your Python example to the proposal.)

Reply:
Looks nice and familiar enough to use easily, that's exciting! Feel free to close the thread.